### PR TITLE
Add `Snowflake ID` to `Bloblang`

### LINKED ISSUE #1377: Feature suggestion: Add snowflake ID to boblang
Hey, Benthos Team.

Recently we come is a situation that we could use `Snowflake ID` in Benthos. 
If it sounds like a good idea, I'd be more than happy to create a PR and add this feature.

### PR DESCRIPTION
**Pull request by ehcaning**
Here is a tiny PR to add `snowflake_id` support to Benthos.
Resolves #1377

### CODE REVIEW

File: internal/bloblang/query/functions.go [hunk 1]
```diff
@@ -699,6 +700,24 @@ var _ = registerSimpleFunction(
...
 //------------------------------------------------------------------------------
 
+var _ = registerSimpleFunction(
+	NewFunctionSpec(
+		FunctionCategoryGeneral, "snowflake_id",
+		"Generate a new snowflake ID each time it is invoked and prints a string representation.",
+		NewExampleSpec("", `root.id = snowflake_id()`),
+	),
+	func(_ FunctionContext) (interface{}, error) {
+		node, err := snowflake.NewNode(1)
```
Comment (mihaitodor) [inline internal/bloblang/query/functions.go:710]: I guess some users might want to pass in a custom node number, but not sure. Maybe consider having it as an optional parameter with `1` as the default value

File: internal/bloblang/query/functions.go [hunk 2]
```diff
@@ -699,6 +700,24 @@ var _ = registerSimpleFunction(
...
+		if err != nil {
+			panic(err)
```
Comment (mihaitodor) [inline internal/bloblang/query/functions.go:712]: We'd rather avoid panicking here if possible. Please have a look at other functions for error handling patterns.

**Review by mihaitodor — CHANGES_REQUESTED** [review]
Thanks for this addition! Since the underlying library is doing some small computations [on init](https://github.com/bwmarrin/snowflake/blob/3cf6d89d1fe90688e862739ec26be74ebef77a57/snowflake.go#L57-L76), it would be better to have this added [here](https://github.com/benthosdev/benthos/blob/main/internal/impl/lang/bloblang.go) instead so people can opt out from importing it when creating a custom binary.

File: internal/bloblang/query/functions.go (hunk 1 above)
Comment (ehcaning) [reply internal/bloblang/query/functions.go]: Thanks for the suggestion, it's done ✅

File: internal/bloblang/query/functions.go (hunk 2 above)
Comment (ehcaning) [reply internal/bloblang/query/functions.go]: Thanks for the suggestion, it's done ✅

**Comment by mihaitodor** [comment]
Thanks for the changes! Don't forget to move the new code to https://github.com/benthosdev/benthos/blob/main/internal/impl/lang/bloblang.go as mentioned above

**Comment by mihaitodor** [comment]
> > Thanks for the changes! Don't forget to move the new code to https://github.com/benthosdev/benthos/blob/main/internal/impl/lang/bloblang.go as mentioned above
> 
> Sorry, I'm a bit lost here, can you explain more please? I don't know what should I do 😄

No prob! You added the functionality to `internal/bloblang/query/functions.go`, but these are core Benthos functions which should come by default for anyone who imports `github.com/benthosdev/benthos/v4/public/components/pure` into their code. However, there's also `github.com/benthosdev/benthos/v4/public/components/pure/extended` for such extras that might not get used frequently and add bloat. In Go, all [`init()` functions](https://go.dev/ref/spec#Package_initialization) get run on startup before `main()`, so, ideally, we'd want to make it possible for people to skip importing them. So, instead of importing `github.com/benthosdev/benthos/v4/public/components/all`, they might cherry-pick only the imports they need from https://github.com/benthosdev/benthos/tree/main/public/components. Given this, you'll want to move your code to https://github.com/benthosdev/benthos/blob/main/internal/impl/lang/bloblang.go and https://github.com/benthosdev/benthos/blob/main/internal/impl/lang/bloblang_test.go so it only gets picked up when people import `github.com/benthosdev/benthos/v4/public/components/pure/extended` (or `github.com/benthosdev/benthos/v4/public/components/all`). Hope that makes sense.

**Comment by ehcaning** [comment]
> > > Thanks for the changes! Don't forget to move the new code to https://github.com/benthosdev/benthos/blob/main/internal/impl/lang/bloblang.go as mentioned above
> > 
> > 
> > Sorry, I'm a bit lost here, can you explain more please? I don't know what should I do 😄
> 
> No prob! You added the functionality to `internal/bloblang/query/functions.go`, but these are core Benthos functions which should come by default for anyone who imports `github.com/benthosdev/benthos/v4/public/components/pure` into their code. However, there's also `github.com/benthosdev/benthos/v4/public/components/pure/extended` for such extras that might not get used frequently and add bloat. In Go, all [`init()` functions](https://go.dev/ref/spec#Package_initialization) get run on startup before `main()`, so, ideally, we'd want to make it possible for people to skip importing them. So, instead of importing `github.com/benthosdev/benthos/v4/public/components/all`, they might cherry-pick only the imports they need from https://github.com/benthosdev/benthos/tree/main/public/components. Given this, you'll want to move your code to https://github.com/benthosdev/benthos/blob/main/internal/impl/lang/bloblang.go and https://github.com/benthosdev/benthos/blob/main/internal/impl/lang/bloblang_test.go so it only gets picked up when people import `github.com/benthosdev/benthos/v4/public/components/pure/extended` (or `github.com/benthosdev/benthos/v4/public/components/all`). Hope that makes sense.

Thanks for the detailed comment. I've updated my PR.
I'll appreciate a review. 🤝

**Review by mihaitodor — APPROVED** [review]
LGTM. Might be nice to place a sample Snowflake ID in the description, so people will know what they're getting when looking at the docs

**Comment by ehcaning** [comment]
@mihaitodor Thanks, can we go ahead and merge it?

**Comment by mihaitodor** [comment]
> @mihaitodor Thanks, can we go ahead and merge it?

Unfortunately, it will have to wait until @Jeffail has a bit of time to look at it, since he's the only maintainer. If it's urgent for you, the best I can suggest is to use a custom build.

**Review by Jeffail — APPROVED** [review]
Looks good, thanks @ehcaning!

**Pull request merged**